lib/spack/spack/package.py

```diff
@@ -2607,27 +2647,32 @@ def _run_default_install_time_test_callbacks(self):
         if self.install_time_test_callbacks is None:
             return
 
-        for name in self.install_time_test_callbacks:
-            try:
-                fn = getattr(self, name)
-            except AttributeError:
-                msg = 'RUN-TESTS: method not implemented [{0}]'
-                tty.warn(msg.format(name))
-            else:
-                tty.msg('RUN-TESTS: install-time tests [{0}]'.format(name))
-                fn()
+        # Log running install-time tests to the build log
+        tty.msg('Running install-time tests')
 
+        # Ensure capture of output for tests that use run_test
+        with self._setup_test():
+            for name in self.install_time_test_callbacks:
+                try:
+                    fn = getattr(self, name)
+                except AttributeError:
+                    msg = 'RUN-TESTS: method not implemented [{0}]'
+                    tty.warn(msg.format(name))
+                else:
+                    tty.msg('RUN-TESTS: install-time tests [{0}]'
+                            .format(name))
+                    try:
+                        fn()
+                    except BaseException as exc:
+                        self.test_failures.append((exc, str(exc)))
 
-def test_process(pkg, kwargs):
-    with tty.log.log_output(pkg.test_log_file) as logger:
-        with logger.force_echo():
-            tty.msg('Testing package {0}'
-                    .format(pkg.test_suite.test_pkg_id(pkg.spec)))
+        # Make sure the build log contains the failure(s)
+        for _, msg in self.test_failures:
+            tty.msg(msg)
```

Comment:
Why is this needed here but not in `test_process`? Having trouble parsing this.

Reply:
The callbacks are performed after the install-phase and, as I mention in the comment at line 2669, this ensures the test messages are reflected in the build log.  

We don't need it in `test_process`, because it is associated with `do_test`, which is tied to the stand-alone test process and uses different logging.  (Note `do_test` is the method used by reporters to capture stand-alone test output.)